### PR TITLE
[opt](function) optimize from_unixtime/date_format by specially format str

### PR DESCRIPTION
**Pull request by Mryange**
## Proposed changes

```
 mysql [test]>select count(date_format(a, 'yyyyMMdd')) from date_format_tmp;
+-----------------------------------+
| count(date_format(a, 'yyyyMMdd')) |
+-----------------------------------+
|                          16000000 |
+-----------------------------------+
1 row in set (0.53 sec)


mysql [test]>select count(date_format(a, 'yyyyMMdd')) from date_format_tmp;
+-----------------------------------+
| count(date_format(a, 'yyyyMMdd')) |
+-----------------------------------+
|                          16000000 |
+-----------------------------------+
1 row in set (0.28 sec)
```

### CODE REVIEW

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run performance

File: be/src/vec/functions/date_format_type.h [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <variant>
+
+#include "vec/common/string_ref.h"
+namespace doris::vectorized {
```
Comment (HappenLee) [inline be/src/vec/functions/date_format_type.h:23]: doris::vectorized::time_format_type

File: be/src/vec/functions/date_format_type.h [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+namespace time_format_type {
+
+template <typename T>
+void inline put_year(T y, char* buf, int& i) {
```
Comment (HappenLee) [inline be/src/vec/functions/date_format_type.h:28]: no need inline, the template func must be inline

File: be/src/vec/functions/function_datetime_string_to_string.h [hunk 3]
```diff
@@ -78,42 +123,65 @@ class FunctionDateTimeStringToString : public IFunction {
...
+        }
+
+        StringRef format(format_state->format_str);
+
+        auto len = ts.size();
+        res_offsets.resize(len);
+        res_data.reserve(len * format.size + len);
+        null_map.resize_fill(len, false);
+
+        if (!format_state->is_valid) {
```
Comment (HappenLee) [inline be/src/vec/functions/function_datetime_string_to_string.h:169]: is_valid should set null, is danger behavioer

**Comment by Mryange** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

**Comment by Mryange** [comment]
run performance

**Review by zhangstar333 — APPROVED** [review]

**Pull request merged**